File: backend/src/spec/api/game_controller_integration_spec.rb
require 'java'
require 'airborne'
require_relative '../spec_helper'
require_relative '../spring_context'

describe '/games post' do

  before(:all) do
    @ctx = SpringContext.instance.spring_ctx
    @port = SpringContext.instance.spring_port
  end
  
  start_game_params = [
      {playerName: "pelaaja", gameDifficulty: "NORMAL"},
      {playerName: "playah", gameDifficulty: "NORMAL"}
  ]

  start_game_params.each do |start|
    player_name = start[:playerName]
    game_difficulty = start[:gameDifficulty]
    context "when creating with playerName as #{player_name} and difficulty as #{game_difficulty}" do

        before(:all) do
          post 'http://localhost:'+@port.to_s+'/games', { },
               { params: { playerName: player_name, difficultyLevel: game_difficulty } }
        end

        let(:gameRepository) { @ctx.getBean "gameRepository" }

        it 'should insert to db' do
          expect(gameRepository.findAll.size).to eq 1
        end

        it 'should return 200 ok' do
          expect_status 200
        end

        it "should set given playerName: #{player_name} for game" do
          expect(json_body[:playerName]).to eq(player_name)
        end

        it "should create game with given #{game_difficulty} difficulty level" do
          expect(json_body[:difficultyLevel]).to eq(game_difficulty)
        end
    end
  end

end

File: backend/src/integration-test/java/fi/aalto/ekanban/controllers/GameControllerIntegrationSpec.java
package fi.aalto.ekanban.controllers;

import static com.greghaskins.spectrum.Spectrum.*;
import static fi.aalto.ekanban.ApplicationConstants.GAME_PATH;
import static io.restassured.RestAssured.given;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.equalTo;

import java.net.HttpURLConnection;
import java.util.function.Supplier;

import com.greghaskins.spectrum.Spectrum;
import io.restassured.RestAssured;
import io.restassured.response.Response;
import org.junit.runner.RunWith;

import fi.aalto.ekanban.enums.GameDifficulty;
import fi.aalto.ekanban.repositories.GameRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.TestContextManager;

@RunWith(Spectrum.class)
@SpringBootTest(webEnvironment= SpringBootTest.WebEnvironment.RANDOM_PORT)
public class GameControllerIntegrationSpec {

    private static final String PLAYER_NAME = "playerName";
    private static final String DIFFICULTY_LEVEL_NAME = "difficultyLevel";

    @Autowired
    private GameRepository gameRepository;

    @Value("${local.server.port}")
    int port;

    {

        beforeAll(() -> {
            new TestContextManager(getClass()).prepareTestInstance(this);
            RestAssured.port = port;
            gameRepository.deleteAll();
        });

        afterAll(() -> {
            gameRepository.deleteAll();
        });

        describe("/games post", () -> {

            describe("when creating", () -> {

                final Supplier<String> playerName = let(() -> "Player");

                describe("with normal difficulty", () -> {

                    final Supplier<GameDifficulty> gameDifficulty = let(() -> GameDifficulty.NORMAL);

                    final Supplier<Response> response = let(() -> given()
                            .formParam(PLAYER_NAME, "Player")
                            .formParam(DIFFICULTY_LEVEL_NAME, GameDifficulty.NORMAL)
                            .when().post(GAME_PATH));

                    it("should return 200 ok", () -> {
                        response.get().then().statusCode(HttpURLConnection.HTTP_OK);
                    });

                    it("should insert game to db", () -> {
                        assertThat(gameRepository.findAll().size(), equalTo(1));
                    });

                    it("should set given playerName for game", () -> {
                        response.get().then().body(PLAYER_NAME, equalTo(playerName.get()));
                    });

                    it("should set given gameDifficulty for game", () -> {
                        response.get().then().body("difficultyLevel", equalTo(gameDifficulty.get().toString()));
                    });

                });
            });

        });
    }

}
